### ISSUE TITLE
tsan address space mapping on NetBSD/amd64 with ASLR

### ISSUE THREAD

**Issue by krytarowski** [google/sanitizers#868]
Hello,

I'm trying to add NetBSD process map in tsan (`lib/tsan/rtl/tsan_platform.h`).

There is ASLR for user-space process map enforced.

The stack grows downards and has preserved layout for 128GB of RAM.
With a memory gap under the stack there is heap that is being allocated top-down.

ld.elf_so maps libraries with mmap(2) on the heap.

```
struct Mapping {
  static const uptr kLoAppMemBeg   = 0x000000000000ull;
  static const uptr kLoAppMemEnd   = 0x300000000000ull;

  static const uptr kShadowBeg     = ;
  static const uptr kShadowEnd     = ;

  static const uptr kMetaShadowBeg = ;
  static const uptr kMetaShadowEnd = ;

  static const uptr kTraceMemBeg   = ;
  static const uptr kTraceMemEnd   = ;

  static const uptr kHeapMemBeg    = 0x600000000000ull;
  static const uptr kHeapMemEnd    = 0x7e0000000000ull;

  static const uptr kHiAppMemBeg   = 0x7e0000000000ull;
  static const uptr kHiAppMemEnd   = 0x7f7ffffff000ull;

  static const uptr kAppMemMsk     = ;
  static const uptr kAppMemXor     = ;
};
```

Is it possible to add shadow, metashadow, trace and appmem mask/xor for so wide address space?

**Comment by krytarowski** [comment]
@dvyukov ^

**Comment by krytarowski** [comment]
I think we could optimize the above mapping.. but I might need some help to consult the proper layout.

**Comment by dvyukov** [comment]
I think such mapping won't work per-se because tsan shadow is 4x of user memory, so there is just not enough virtual address space to fit shadow for all of these ranges.
However, the heap should be 16x smaller, and can be placed at 0x7d0000000000-0x7e0000000000. And if we use "32-bit" allocator, then we can combine heap and the high range (where mmap's are placed). The 32-bit allocator does not need reserved range, it just uses normal mmap's to allocate memory.
The low range is still quite large. I don't have good ideas as to what to do with it... Would it be feasible to reserve a smaller range and then re-exec the process until kernel chooses to place binary/stack (or whatever is situated in the low range) where are want them to be?

**Comment by dvyukov** [comment]
> I think we could optimize the above mapping.. but I might need some help to consult the proper layout.

Mapping and layout are tightly tied together. They need to be built at the same time. There is no general recipe for this, only creativity.

**Comment by krytarowski** [comment]
Thank you for your answers. My feeling was like hitting a wall and I needed to consult the current state with more knowledgeable people.

> I think such mapping won't work per-se because tsan shadow is 4x of user memory, so there is just not enough virtual address space to fit shadow for all of these ranges.

Right. This is what worries me.

I've uploaded `pmap`(1) output (`cat /proc/*/maps` equivalent) for Firefox and Thunderbird running for a long day and doing real job (not just started):

http://netbsd.org/~kamil/llvm/firefox.mapping.aslr.txt
http://netbsd.org/~kamil/llvm/thunderbird.mapping.aslr.txt

Executables are without PIE, as with PIE they can go wild and ASLR will randomize the layout of the `.text` section.

NetBSD supports PIE binaries, but they are disabled by default right now. For PIE binaries we aren't using the Linux middle-app area `0x5500 0000 0000`-`0x5600 0000 0000`, but we are shifting the .text section from the defined address in an ELF file (somewhere near `0x0` or more technically in modern executables `0x1000`).

There is something odd that I don't understand right now that something preallocates memory at the 0x1000 0000 0000 - 0x3000 0000 0000 range in Firefox and Thunderbird. Maybe this is firefox/thunderbird specific that they request these addresses.. I'm trying to investigate it. I'm not reproducing it so far for other programs.

There is an option to support tsan with ASLR disabled mode only and this should make the general job easier. It's sufficient to attach an ELF note to disable PaX ASLR per-program.

I think that NetBSD can reuse the Linux/FreeBSD mapping for `kLoAppMemBeg`-`kLoAppMemEnd`.

```
0000 0000 1000 - 0080 0000 0000: main binary
```

> However, the heap should be 16x smaller, and can be placed at 0x7d0000000000-0x7e0000000000. And if we use "32-bit" allocator, then we can combine heap and the high range (where mmap's are placed). The 32-bit allocator does not need reserved range, it just uses normal mmap's to allocate memory.

What do you mean by 32-bit allocator? `MAP_32BIT`? If so, than `MAP_32BIT` isn't supported on NetBSD and no need to worry about it.

I've disabled globally ASLR (`sysctl`(8) switch) and here is a mapping of `cat`(1):

```
$ cat /proc/self/maps 
0000000000200000-0000000000202000 r-xp 0000000000000000 00:00 8503059                            /bin/cat
0000000000402000-0000000000403000 r--p 0000000000002000 00:00 8503059                            /bin/cat
0000000000403000-0000000000404000 rw-p 0000000000000000 00:00 0                                  
00007f7ff7400000-00007f7ff7403000 r-xp 0000000000000000 00:00 11356967                           /usr/lib/i18n/libUTF8.so.5.0
00007f7ff7403000-00007f7ff7602000 ---p 0000000000003000 00:00 11356967                           /usr/lib/i18n/libUTF8.so.5.0
00007f7ff7602000-00007f7ff7603000 rw-p 0000000000002000 00:00 11356967                           /usr/lib/i18n/libUTF8.so.5.0
00007f7ff7700000-00007f7ff7710000 rw-p 0000000000000000 00:00 0                                  
00007f7ff7710000-00007f7ff7800000 rw-p 0000000000000000 00:00 0                                  
00007f7ff7800000-00007f7ff7952000 r-xp 0000000000000000 00:00 12381666                           /lib/libc.so.12.208
00007f7ff7952000-00007f7ff7b51000 ---p 0000000000152000 00:00 12381666                           /lib/libc.so.12.208
00007f7ff7b51000-00007f7ff7b58000 r--p 0000000000151000 00:00 12381666                           /lib/libc.so.12.208
00007f7ff7b58000-00007f7ff7b5e000 rw-p 0000000000158000 00:00 12381666                           /lib/libc.so.12.208
00007f7ff7b5e000-00007f7ff7b6f000 rw-p 0000000000000000 00:00 0                                  
00007f7ff7c00000-00007f7ff7c10000 r-xp 0000000000000000 00:00 10278417                           /libexec/ld.elf_so
00007f7ff7c10000-00007f7ff7e10000 ---p 0000000000000000 00:00 0                                  
00007f7ff7e10000-00007f7ff7e11000 rw-p 0000000000010000 00:00 10278417                           /libexec/ld.elf_so
00007f7ff7e11000-00007f7ff7e12000 rw-p 0000000000000000 00:00 0                                  
00007f7ff7ef3000-00007f7ff7eff000 rw-p 0000000000000000 00:00 0                                  
00007f7ff7eff000-00007f7fffbff000 ---p 0000000000000000 00:00 0                                  
00007f7fffbff000-00007f7fffff0000 rw-p 0000000000000000 00:00 0                                  
00007f7fffff0000-00007f7ffffff000 rw-p 0000000000000000 00:00 0 
```

(always repeatable result)

> The low range is still quite large. I don't have good ideas as to what to do with it... Would it be feasible to reserve a smaller range and then re-exec the process until kernel chooses to place binary/stack (or whatever is situated in the low range) where are want them to be?

I think we can assume now `0000 0000 1000` - `0080 0000 0000`.

I'm still investigating the `0x1000 0000 0000` - `0x3000 0000 0000` allocations in Firefox and Thunderbird, so far I haven't seen them in other programs.

I'm attaching a LibreOffice Calc executable running from Linux binary on NetBSD natively:

http://netbsd.org/~kamil/llvm/scalc.mapping.aslr.txt

> Mapping and layout are tightly tied together. They need to be built at the same time. There is no general recipe for this, only creativity.

I don't understand this kAppMemXor trick...

How would you merge `kHiAppMem` and `kHeapMem`?

Do you think this new mapping makes sense:

```
struct Mapping {
  static const uptr kLoAppMemBeg   = 0x000000000000ull;
  static const uptr kLoAppMemEnd   = 0x008000000000ull;

  static const uptr kShadowBeg     = 0x010000000000ull;
  static const uptr kShadowEnd     = 0x200000000000ull;

  static const uptr kMetaShadowBeg = 0x300000000000ull;
  static const uptr kMetaShadowEnd = 0x340000000000ull;

  static const uptr kTraceMemBeg   = 0x600000000000ull;
  static const uptr kTraceMemEnd   = 0x620000000000ull;

  static const uptr kHeapMemBeg    = 0x7d0000000000ull;
  static const uptr kHeapMemEnd    = 0x7e0000000000ull;

  static const uptr kHiAppMemBeg   = 0x7e0000000000ull;
  static const uptr kHiAppMemEnd   = 0x7f7ffffff000ull;

  static const uptr kAppMemMsk     = 0x780000000000ull; /// ???
  static const uptr kAppMemXor     = 0x040000000000ull; /// ???
};
```

Correction, the stack size on NetBSD/amd64 is 128MB, not GB.

**Comment by krytarowski** [comment]
Ping! With feedback I will scratch something over the weekend.

**Comment by dvyukov** [comment]
> There is something odd that I don't understand right now that something preallocates memory at the 0x1000 0000 0000 - 0x3000 0000 0000 range in Firefox and Thunderbird. Maybe this is firefox/thunderbird specific that they request these addresses.

Does it present on a "hello, world" binary?

> There is an option to support tsan with ASLR disabled mode only and this should make the general job easier. It's sufficient to attach an ELF note to disable PaX ASLR per-program.

Sounds good. We don't need ASLR with tsan. And we can attach a note with inline asm, right?

> What do you mean by 32-bit allocator? MAP_32BIT? If so, than MAP_32BIT isn't supported on NetBSD and no need to worry about it.

It's a version of tsan internal heap allocator. See SizeClassAllocator32 in tsan_rtl.h. This allocator does not need a huge reserved range, instead it just calls mmap and uses whatever memory OS returns.

> I've disabled globally ASLR (sysctl(8) switch) and here is a mapping of cat(1):

It seems we can just reuse linux mapping for this without any changes.

> Do you think this new mapping makes sense:

Let's try to reuse linux mapping. That would be the simplest option.

> I don't understand this kAppMemXor trick... How would you merge kHiAppMem and kHeapMem?

Tsan linux mapping allows 4 1TB user memory ranges: one is around zero (non-pie binaries, 43-bit mmaps), one is around 0x55 (pie binaries with aslr), one is around 0x7b (heap), and one is around 0x7f (mmaps and pie binaries on older versions). kAppMemXor mutiplexes these ranges onto 4 4TB shadow ranges in [kShadowBeg:kShadowEnd).
The tricky part is that the first 1/4 of shadow range can't be used, because it is around 0 addresses and overlaps with non-pie binary. So the 4 user ranges are actually multiplexed on 3 shadow ranges (the first one is unused). The trick is that we can have either 0x7f or 0x55, but not both, so they are mapped onto the same shadow range.

**Comment by krytarowski** [comment]
> Does it present on a "hello, world" binary?

No, this 0x1000 0000 0000 - 0x3000 0000 0000 range is only Firefox+Thunderbird. We can skip now FF+TB.
> Sounds good. We don't need ASLR with tsan. And we can attach a note with inline asm, right?

Yes, it's possible to attach inline asm note. Certainly the same will apply for msan.

> It's a version of tsan internal heap allocator. See SizeClassAllocator32 in tsan_rtl.h. This allocator does not need a huge reserved range, instead it just calls mmap and uses whatever memory OS returns.

I will have a look.

> It seems we can just reuse linux mapping for this without any changes.

If I recall correctly it's not that simple, because we might land between heap and hiAppMem. But I will make sure with ASLR disabled globally. If it will work I will create a dedicated ELF note.

```
linux/freebsd amd64
0000 0000 1000 - 0080 0000 0000: main binary and/or MAP_32BIT mappings (512GB)
0040 0000 0000 - 0100 0000 0000: -
0100 0000 0000 - 2000 0000 0000: shadow
2000 0000 0000 - 3000 0000 0000: -
3000 0000 0000 - 4000 0000 0000: metainfo (memory blocks and sync objects)
4000 0000 0000 - 5500 0000 0000: -
5500 0000 0000 - 5680 0000 0000: pie binaries without ASLR or on 4.1+ kernels
5680 0000 0000 - 6000 0000 0000: -
6000 0000 0000 - 6200 0000 0000: traces
6200 0000 0000 - 7d00 0000 0000: -
7b00 0000 0000 - 7c00 0000 0000: heap
7c00 0000 0000 - 7e80 0000 0000: -
7e80 0000 0000 - 8000 0000 0000: modules and main thread stack
```

> Tsan linux mapping allows 4 1TB user memory ranges: one is around zero (non-pie binaries, 43-bit mmaps), one is around 0x55 (pie binaries with aslr), one is around 0x7b (heap), and one is around 0x7f (mmaps and pie binaries on older versions). kAppMemXor mutiplexes these ranges onto 4 4TB shadow ranges in [kShadowBeg:kShadowEnd).
The tricky part is that the first 1/4 of shadow range can't be used, because it is around 0 addresses and overlaps with non-pie binary. So the 4 user ranges are actually multiplexed on 3 shadow ranges (the first one is unused). The trick is that we can have either 0x7f or 0x55, but not both, so they are mapped onto the same shadow range.

Thanks, I will research it. At least for the curiosity.

**Comment by krytarowski** [comment]
OK, this mapping works with ASLR disabled.

I'm not struggling with libpthread init code, it uses internally mmap(2), it's being intercepted by tsan an.. and tsan attempts to create an internal thread before `pthread__init` accomplishes.

This leads to internal assert that generates SIGABRT.

```
#0  0x00007f7ff69103ba in _lwp_kill () from /usr/lib/libc.so.12
#1  0x0000000000418852 in __interceptor_raise () at /public/llvm/projects/compiler-rt/lib/tsan/rtl/tsan_interceptors.cc:1989
#2  0x00007f7ff683ec7d in __libc_thr_create_stub (tp=tp@entry=0x7f7fffffe480, ta=ta@entry=0x0, f=f@entry=0x44c430 <BackgroundThread()>, a=a@entry=0x0)
    at /usr/src/lib/libc/thread-stub/thread-stub.c:421
#3  0x00007f7ff700b05a in pthread_create (thread=0x7f7fffffe480, attr=0x0, startfunc=0x44c430 <BackgroundThread()>, arg=0x0)
    at /usr/src/lib/libpthread/pthread.c:426
#4  0x000000000045cdb1 in internal_start_thread () at /public/llvm/projects/compiler-rt/lib/sanitizer_common/sanitizer_linux.cc:1672
#5  0x000000000044450a in StartBackgroundThread () at /public/llvm/projects/compiler-rt/lib/tsan/rtl/tsan_rtl.cc:226
#6  Initialize () at /public/llvm/projects/compiler-rt/lib/tsan/rtl/tsan_rtl.cc:375
#7  0x000000000040ff64 in ScopedInterceptor () at /public/llvm/projects/compiler-rt/lib/tsan/rtl/tsan_interceptors.cc:267
#8  0x0000000000412250 in __interceptor_mmap () at /public/llvm/projects/compiler-rt/lib/tsan/rtl/tsan_interceptors.cc:709
#9  0x00007f7ff7007760 in pthread_tsd_init (tlen=tlen@entry=0x7f7ff72117c0) at /usr/src/lib/libpthread/pthread_tsd.c:88
#10 0x00007f7ff700bf88 in pthread__init () at /usr/src/lib/libpthread/pthread.c:183
#11 0x00007f7ff6911215 in _libc_init () at /usr/src/lib/libc/misc/initfini.c:128
#12 0x00007f7ff683afa4 in ?? () from /usr/lib/libc.so.12
#13 0x00007f7ff7ef83a0 in ?? ()
#14 0x00007f7ff68363b9 in _init () from /usr/lib/libc.so.12
#15 0x0000000000000000 in ?? ()
```

I'm trying to investigate.

**Comment by krytarowski** [comment]
http://netbsd.org/~kamil/llvm/tsan.2017-10-14.txt I got the sanitizers to work for the first time ever!

**Comment by dvyukov** [comment]
Great! What was the fix for the pthread__init issue?

**Comment by krytarowski** [comment]
In general tsan isn't reliable.. I managed to hack it to work for ~5% attempts of execution for the same program.

The underlying problem is starting internal background thread for the first intercepted call. This is not possible on NetBSD, at least with the pthread(3) API. We are requested to finish libc and libpthread startup code - it's started and finished before executing constructors from the code.

Possible solutions:
 - switch to low-level _lwp_create(2) calls to manually create threads without libc/libpthread
 - intercept main() and in main() start the background thread.

The _lwp_create(2) approach should not be used unless no other choice and it does not guarantee that it will keep not failing for the same reason as pthread(3) but for other calls.

Can we defer this code to main(), if so can we do it for all OS'es? C, C++, golong use `main()`. Shutting into the dark, this might unbreak FreeBSD as well (but I don't have access to a FreeBSD machine capable to build llvm projects and test it myself).

```
diff --git a/lib/tsan/rtl/tsan_rtl.cc b/lib/tsan/rtl/tsan_rtl.cc
index 882e06765..56586ca7c 100644
--- a/lib/tsan/rtl/tsan_rtl.cc
+++ b/lib/tsan/rtl/tsan_rtl.cc
@@ -371,7 +371,7 @@ void Initialize(ThreadState *thr) {
   // On MIPS, TSan initialization is run before
   // __pthread_initialize_minimal_internal() is finished, so we can not spawn
   // new threads.
-#ifndef __mips__
+#if !defined(__mips__) && !SANITIZER_NETBSD
   StartBackgroundThread();
   SetSandboxingCallback(StopBackgroundThread);
 #endif
```

As a work around I have disabled it at all temporarily. As far as I understand it's doing G/C.

I also got a local hack:

```
@@ -1618,10 +1667,10 @@ void *internal_start_thread(void(*func)(void *arg), void *arg) {
   // on any thread, setuid call hangs (see test/tsan/setuid.c).
   internal_sigdelset(&set, 33);
 #endif
-  internal_sigprocmask(SIG_SETMASK, &set, &old);
+//  internal_sigprocmask(SIG_SETMASK, &set, &old);
   void *th;
   real_pthread_create(&th, nullptr, (void*(*)(void *arg))func, arg);
-  internal_sigprocmask(SIG_SETMASK, &old, nullptr);
+//  internal_sigprocmask(SIG_SETMASK, &old, nullptr);
   return th;
 }
```

This is wrong but it enabled me to get tsan to work sometimes.

I'm getting right now without the above hack for `internal_sigprocmask()`:

```
$ ./a.out                                                                                                                    
assertion "pthread__tsd_destructors[key] != NULL" failed: file "/usr/src/lib/libpthread/pthread_tsd.c", line 176, function "pthread__add_specific"
Abort
```

I'm working on it.

**Comment by dvyukov** [comment]
The background thread is not critical for tsan operation. It flushes shadow memory periodically and it can also write heap profile periodically. Let's proceed with `#if !defined(__mips__) && !SANITIZER_NETBSD` for now. We can make all tests pass without the background thread and tsan already will be useful for real programs. I am not sure if we can intercept main. Maybe we could start it when the program starts it's first thread (second after main), at that point we know that threads are working. But we will need to do some additional testing for such changes.

**Comment by krytarowski** [comment]
Thank you for the feedback! This sounds good.

I'm investigating and in case of doubts I will raise more questions.

**Comment by krytarowski** [comment]
There is another similar issue in `__tsan::InitializeInterceptors()`. I've disabled it locally.

```
#if !SANITIZER_MAC
  if (pthread_key_create(&g_thread_finalize_key, &thread_finalize)) {
    Printf("ThreadSanitizer: failed to create thread key\n");
    Die();
  }
#endif
```

My current results for check-sanitizer.

```
Expected Passes    : 148                    
Expected Failures  : 1
Unsupported Tests  : 82
Unexpected Failures: 66
```

I note two issues:
* atexit(3) or similar function breaks, I keep locally a patch:


```
index 82a7f371a..6b6a2233b 100644
--- a/lib/tsan/rtl/tsan_interceptors.cc
+++ b/lib/tsan/rtl/tsan_interceptors.cc
@@ -387,9 +387,9 @@ static void at_exit_wrapper(void *arg) {
   ThreadState *thr = cur_thread();
   uptr pc = 0;
   Acquire(thr, pc, (uptr)arg);
-  AtExitCtx *ctx = (AtExitCtx*)arg;
-  ((void(*)(void *arg))ctx->f)(ctx->arg);
-  InternalFree(ctx);
+//  AtExitCtx *ctx = (AtExitCtx*)arg;
+//  ((void(*)(void *arg))ctx->f)(ctx->arg);
+//  InternalFree(ctx);
 }
 
 static int setup_at_exit_wrapper(ThreadState *thr, uptr pc, void(*f)(),
```

* inappropriate thread state change from Running to Dead, instead of
from Running to Finished to Dead

```
--- a/lib/sanitizer_common/sanitizer_thread_registry.cc
+++ b/lib/sanitizer_common/sanitizer_thread_registry.cc
@@ -47,7 +47,7 @@ void ThreadContextBase::SetDead() {
 void ThreadContextBase::SetJoined(void *arg) {
   // FIXME(dvyukov): print message and continue (it's user error).
   CHECK_EQ(false, detached);
-  CHECK_EQ(ThreadStatusFinished, status);
+//  CHECK_EQ(ThreadStatusFinished, status);
   status = ThreadStatusDead;
   user_id = 0;
   OnJoined(arg);
```

I'm investigating whether this is the result of disabled `pthread_key_create(&g_thread_finalize_key, &thread_finalize)`

**Comment by krytarowski** [comment]
The same with premature call of `__cxa_atexit()`:

```
  if (REAL(__cxa_atexit)(&finalize, 0, 0)) {
    Printf("ThreadSanitizer: failed to setup atexit callback\n");
    Die();
  }
```

**Comment by krytarowski** [comment]
Hmm, are there any suggestions how/when to defer these calls?

I'm experimenting with constructors etc, but so far nothing is a step forwards.

**Comment by dvyukov** [comment]
Which calls exactly?
What I meant is starting the background call in the first pthread_create interceptor.
Or you mean pthread_key_create/__cxa_atexit? Maybe we could do initialization only in preinit array hook, and before that just forward all interceptors to REAL versions? So that mmap interceptor coming from pthread initialization does not actually do anything. However, this would require initializing at least REAL pointers with dlsym.

**Comment by krytarowski** [comment]
I mean pthread_key_create/_cxa_atexit. They break internally as there are no internal structs initialized etc.

I will give it a try to initialize them with REAL versions.

**Comment by krytarowski** [comment]
For the record, check-tsan as of today on NetBSD/amd64:

```
Expected Passes    : 226 
Expected Failures  : 1
Unsupported Tests  : 82
Unexpected Failures: 67
```

**Comment by krytarowski** [comment]
I think I got some success with deferring the interception calls.

`check-tsan`:

```
Expected Passes    : 245 (+19)
Expected Failures  : 1
Unsupported Tests  : 82
Unexpected Failures: 48 (-19)
```

**Comment by krytarowski** [comment]
I've pushed ready local patches as https://reviews.llvm.org/D39124

With few additional kludges I can reach the following result:

```
    check-tsan:
    
      Expected Passes    : 248
      Expected Failures  : 1
      Unsupported Tests  : 83
      Unexpected Failures: 44
```

I think I hit the wall again with crashes and I will try to test my local compiler-rt changes on a remote Linux host.

**Comment by krytarowski** [comment]
We are down to the issue that certain functions must be deferred to be executed later, after finishing the initialization of libpthread and libc.

List of them:

 - `pthread_key_create()`
 - `__cxa_atexit()`
- `StartBackgroundThread()`

I was trying to call them from a function with a constructor property, but it resulted with anomalies and random NULL pointer dereferences.

```
$NetBSD$

--- lib/tsan/rtl/tsan_interceptors.cc.orig	2017-10-25 19:03:27.000000000 +0000
+++ lib/tsan/rtl/tsan_interceptors.cc
@@ -395,9 +395,9 @@ static void at_exit_wrapper(void *arg) {
   ThreadState *thr = cur_thread();
   uptr pc = 0;
   Acquire(thr, pc, (uptr)arg);
-  AtExitCtx *ctx = (AtExitCtx*)arg;
-  ((void(*)(void *arg))ctx->f)(ctx->arg);
-  InternalFree(ctx);
+//  AtExitCtx *ctx = (AtExitCtx*)arg;
+//  ((void(*)(void *arg))ctx->f)(ctx->arg);
+//  InternalFree(ctx);
 }
 
 static int setup_at_exit_wrapper(ThreadState *thr, uptr pc, void(*f)(),
@@ -920,6 +920,39 @@ extern "C" void *__tsan_thread_start_fun
   return res;
 }
 
+static void FlushStreams() {
+  // Flushing all the streams here may freeze the process if a child thread is
+  // performing file stream operations at the same time.
+
+  REAL(fflush)(stdout);
+  REAL(fflush)(stderr);
+}
+
+static void finalize(void *arg) {
+  using namespace __tsan;
+  ThreadState *thr = cur_thread();
+  int status = Finalize(thr);
+  // Make sure the output is not lost.
+  FlushStreams();
+  if (status)
+    Die();
+}
+
+__attribute__((constructor))
+//__attribute__((section(".init_array")))
+static void
+yikes1(void) {
+  if (REAL(__cxa_atexit)(&finalize, 0, 0)) {
+    Printf("ThreadSanitizer: failed to setup atexit callback\n");
+    Die();
+  }
+
+  if (pthread_key_create(&g_thread_finalize_key, &thread_finalize)) {
+    Printf("ThreadSanitizer: failed to create thread key\n");
+    Die();
+  }
+}
+
 TSAN_INTERCEPTOR(int, pthread_create,
     void *th, void *attr, void *(*callback)(void*), void * param) {
   SCOPED_INTERCEPTOR_RAW(pthread_create, th, attr, callback, param);
@@ -1675,12 +1708,12 @@ TSAN_INTERCEPTOR(void*, tmpfile64, int f
 #define TSAN_MAYBE_INTERCEPT_TMPFILE64
 #endif
 
-static void FlushStreams() {
-  // Flushing all the streams here may freeze the process if a child thread is
-  // performing file stream operations at the same time.
-  REAL(fflush)(stdout);
-  REAL(fflush)(stderr);
-}
+//static void FlushStreams() {
+//  // Flushing all the streams here may freeze the process if a child thread is
+//  // performing file stream operations at the same time.
+//  REAL(fflush)(stdout);
+//  REAL(fflush)(stderr);
+//}
 
 TSAN_INTERCEPTOR(void, abort, int fake) {
   SCOPED_TSAN_INTERCEPTOR(abort, fake);
@@ -2619,16 +2652,16 @@ void InitializeInterceptors() {
   REAL(atexit) = (int(*)(void(*)()))unreachable;
 #endif
 
-  if (REAL(__cxa_atexit)(&finalize, 0, 0)) {
-    Printf("ThreadSanitizer: failed to setup atexit callback\n");
-    Die();
-  }
+//  if (REAL(__cxa_atexit)(&finalize, 0, 0)) {
+//    Printf("ThreadSanitizer: failed to setup atexit callback\n");
+//    Die();
+//  }
 
 #if !SANITIZER_MAC
-  if (pthread_key_create(&g_thread_finalize_key, &thread_finalize)) {
-    Printf("ThreadSanitizer: failed to create thread key\n");
-    Die();
-  }
+//  if (pthread_key_create(&g_thread_finalize_key, &thread_finalize)) {
+//    Printf("ThreadSanitizer: failed to create thread key\n");
+//    Die();
+//  }
 #endif
 
   FdInit();
$NetBSD$

--- lib/tsan/rtl/tsan_rtl.cc.orig	2017-10-17 18:35:43.000000000 +0000
+++ lib/tsan/rtl/tsan_rtl.cc
@@ -371,7 +371,7 @@ void Initialize(ThreadState *thr) {
   // On MIPS, TSan initialization is run before
   // __pthread_initialize_minimal_internal() is finished, so we can not spawn
   // new threads.
-#ifndef __mips__
+#if !defined(__mips__) && !SANITIZER_NETBSD
   StartBackgroundThread();
   SetSandboxingCallback(StopBackgroundThread);
 #endif
```

I need help here to find out a reliable way, that could be reused on Linux to retain the same algorithm+behavior, how to defer these calls. I bet this can also help FreeBSD as they suffer from similar breakage in initialization.

My null-pointer dereference issues (or part of them) might be unrelated and caused by other bugs, but first we need to defer these initialization calls.

**Comment by dvyukov** [comment]
I did not yet have time to think through this. And I will be OOO for a week starting from tomorrow.
Does anything of this also affect asan? If yes, it may be simpler to start with asan (it's generally).

**Comment by krytarowski** [comment]
asan works correctly. There are 1-3 failing tests in asan, one with massive number of threads and the other two are indeterministic and related to mmap(). For now I skip the asan issues as tsan is prioritized, and the issues seem to be unrelated.

I have got another idea how to address it and I will keep experimenting.

**Comment by krytarowski** [comment]
One of the problems in tsan has been investigated.

tsan uses a hack to track whether a thread is exiting. It attached to the on-pthread-destructor deallocation of thread structs. While after calling the destructor the NetBSD's implementation still calls mutex_enter/exit-like functions for a while. I have switched the deallocation to _lwp_exit() and it seams to work better.

It also means that we might have no need to defer initialization calls.

There is another problem (again NULL pointer dereference) with at_exit_wrapper(), I'm investigating it.

**Comment by dvyukov** [comment]
Great. Deferring tsan thread exit callback to a later point (when nothing happens on the thread anymore) looks like the right solution. AFAIR we already use a different thread exit notification mechanism for mac. It just happened that pthread_specific destructor work for linux (and there are no better alternatives).

**Comment by krytarowski** [comment]
I've narrowed down another bug. I'm not sure whether this is NetBSD mistake or tsan. We cannot register __cxa_atexit(a,b,NULL) on NetBSD. We treat it as atexit(a) and drop the 'b' argument.

https://nxr.netbsd.org/xref/src/lib/libc/stdlib/atexit.c#208

I'm test building new libc and distribution to test my change..

```
Index: atexit.c
===================================================================
RCS file: /cvsroot/src/lib/libc/stdlib/atexit.c,v
retrieving revision 1.29
diff -u -r1.29 atexit.c
--- atexit.c    19 Apr 2015 18:15:26 -0000      1.29
+++ atexit.c    2 Nov 2017 17:19:22 -0000
@@ -205,7 +205,7 @@
                if (dso == NULL || dso == ah->ah_dso || ah->ah_atexit == NULL) {
                        if (ah->ah_atexit != NULL) {
                                void *p = atexit_handler_stack;
-                               if (ah->ah_dso != NULL) {
+                               if (ah->ah_dso != NULL || ah->ah_arg != NULL) {
                                        cxa_func = ah->ah_cxa_atexit;
                                        ah->ah_cxa_atexit = NULL;
                                        (*cxa_func)(ah->ah_arg);
```

**Comment by krytarowski** [comment]
I will move the discussion here about the remaining bugs.

The problem of Joined/Detached thread has been analyzed, however I need feedback in order to develop the proper solution.

There are two threads: A and B. A quits, B waits for A to be finished.

Thread's A termination triggers `DestroyThreadState()` (via pthread destructor on Linux or `_lwp_exit()` on NetBSD).

The B thread is notified about the termination of A, just before executing `_lwp_exit()` in A.

A calls `DestroyThreadState()`, B calls `ThreadJoin()`. And these calls race.

The correct order is to call in A: `DestroyThreadState()` and finish its execution; and afterwards start calling `ThreadJoin()` in B.

`_lwp_exit()` is the only way to detect thread's termination on NetBSD and it must stay.

Code - I have added in comments where we need to wait and where we need to signal the end of DestroyThreadState() execution.

```
TSAN_INTERCEPTOR(int, pthread_join, void *th, void **ret) {
  SCOPED_INTERCEPTOR_RAW(pthread_join, th, ret);
  int tid = ThreadTid(thr, pc, (uptr)th);
  ThreadIgnoreBegin(thr, pc);
  int res = BLOCK_REAL(pthread_join)(th, ret);
  ThreadIgnoreEnd(thr, pc);
  if (res == 0) {
    // Wait for tid to finish execution of DestroyThreadState()
    ThreadJoin(thr, pc, tid);
  }
  return res;
} 
```

```
void DestroyThreadState() {
  ThreadState *thr = cur_thread();
  Processor *proc = thr->proc();
  ThreadFinish(thr);
  ProcUnwire(proc, thr);
  ProcDestroy(proc);
  ThreadSignalContext *sctx = thr->signal_ctx;
  if (sctx) {
    thr->signal_ctx = 0;
    UnmapOrDie(sctx, sizeof(*sctx));
  }
  DTLS_Destroy();
  cur_thread_finalize();
  // Notify finish of execution of DestroyThreadState()
}
```

How to develop properly the signal/wait mechanism? Can I access `ThreadState` of A in B on the TSan level?

The exact same problem is shared with LSan.

**Comment by dvyukov** [comment]
> How to develop properly the signal/wait mechanism?

I think we can use a mechanism similar to what is used to synchronize pthread_create/__tsan_thread_start_func. There is basically a spin wait with yield. But since we know the thread has already started exiting, such inefficient wait should not be a problem.

> Can I access ThreadState of A in B on the TSan level?

Yes, sure.
For pthread_create/__tsan_thread_start_func it's a little bit simpler, because we pass explicit context from pthread_create to start func.
I think you can make ThreadTid return ThreadContext instead of just tid. ThreadContext contains both tid and ThreadState.

**Comment by dvyukov** [comment]
What's the exact race scenario with pthread_detach?

**Comment by krytarowski** [comment]
> I think you can make ThreadTid return ThreadContext instead of just tid. ThreadContext contains both tid and ThreadState.

I will give it a try.

> What's the exact race scenario with pthread_detach?

Hmm, for pthread_detach(3) there is no pthread_join(3) so there is no race. Am I missing something?

Current test-results (after a local workaround for this exit/join race):

```
Failing Tests (37):
    ThreadSanitizer-Unit :: unit/./TsanUnitTest-x86_64-Test/Mutex.SpinWrite
    ThreadSanitizer-Unit :: unit/./TsanUnitTest-x86_64-Test/Mutex.Write
    ThreadSanitizer-x86_64 :: bench_mutex.cc
    ThreadSanitizer-x86_64 :: cond.c
    ThreadSanitizer-x86_64 :: cond_cancel.c
    ThreadSanitizer-x86_64 :: cond_destruction.cc
    ThreadSanitizer-x86_64 :: cond_race.cc
    ThreadSanitizer-x86_64 :: cond_version.c
    ThreadSanitizer-x86_64 :: deadlock_detector_stress_test.cc
    ThreadSanitizer-x86_64 :: dtls.c
    ThreadSanitizer-x86_64 :: exceptions.cc
    ThreadSanitizer-x86_64 :: ignore_lib5.cc
    ThreadSanitizer-x86_64 :: ignored-interceptors-mmap.cc
    ThreadSanitizer-x86_64 :: longjmp3.cc
    ThreadSanitizer-x86_64 :: longjmp4.cc
    ThreadSanitizer-x86_64 :: mutex_cycle2.c
    ThreadSanitizer-x86_64 :: mutex_cycle_long.c
    ThreadSanitizer-x86_64 :: mutex_destroy_locked.cc
    ThreadSanitizer-x86_64 :: mutex_lock_destroyed.cc
    ThreadSanitizer-x86_64 :: mutexset1.cc
    ThreadSanitizer-x86_64 :: mutexset2.cc
    ThreadSanitizer-x86_64 :: mutexset3.cc
    ThreadSanitizer-x86_64 :: mutexset4.cc
    ThreadSanitizer-x86_64 :: mutexset5.cc
    ThreadSanitizer-x86_64 :: mutexset6.cc
    ThreadSanitizer-x86_64 :: mutexset7.cc
    ThreadSanitizer-x86_64 :: mutexset8.cc
    ThreadSanitizer-x86_64 :: race_on_mutex.c
    ThreadSanitizer-x86_64 :: race_on_mutex2.c
    ThreadSanitizer-x86_64 :: signal_cond.cc
    ThreadSanitizer-x86_64 :: signal_errno.cc
    ThreadSanitizer-x86_64 :: signal_malloc.cc
    ThreadSanitizer-x86_64 :: signal_sync.cc
    ThreadSanitizer-x86_64 :: signal_sync2.cc
    ThreadSanitizer-x86_64 :: signal_thread.cc
    ThreadSanitizer-x86_64 :: vfork.cc
    ThreadSanitizer-x86_64 :: vptr_benign_race.cc

  Expected Passes    : 221
  Expected Failures  : 1
  Unsupported Tests  : 83
  Unexpected Failures: 37
```

**Comment by krytarowski** [comment]
Regarding mutexes, I still don't understand the problem. It looks like a mutex is registerd&deregistered before notifying the problem.. so the race is detected out of the context of mutex that is being held.

dtls.c has a problem unrelated to TLS, as dlopen(3) + dlsym(3) breaks (NULL received from dlsym(3)).

longjmp(3)/setjmp(3) isn't ported.

**Comment by dvyukov** [comment]
> Hmm, for pthread_detach(3) there is no pthread_join(3) so there is no race. Am I missing something?

I don't know, you mentioned pthread_detach :) If there is no known problem let's leave it aside for now.

**Comment by krytarowski** [comment]
Thank you for the answers for the exit/join issue, I'm going to research a proper patch.

**Comment by krytarowski** [comment]
> I don't know, you mentioned pthread_detach :) If there is no known problem let's leave it aside for now.

Ah right. I have used term detach in the meaning of thread's termination (_lwp_exit(3)). Not pthread_detach(3). Excuse me for wording.

**Comment by dvyukov** [comment]
> Regarding mutexes, I still don't understand the problem. It looks like a mutex is registerd&deregistered before notifying the problem.. so the race is detected out of the context of mutex that is being held.

I am still puzzled and don't have any suggestions. That code should platform-independent. I would start with adding lots of debug printfs to mutexset code. Do we call mutexset functions? Does size even become >0? Is mutexset empty during race reporting? If it's not empty why we don't them to report? etc.

**Comment by krytarowski** [comment]
> I would start with adding lots of debug printfs to mutexset code

I did it a while ago, however I've removed all the debug code locally in order to fix lower-hanging fruits.

I've suspected that maybe TLS handling was culprit, but after getting it to work (still without perfecting it) the problem is still there.

> Do we call mutexset functions?

Yes. We call Add(), Remove(), Add(), Remove(). And when we report any issue, mutexset is always 0.


> Does size even become >0?

Yes, however it's never >1.

> Is mutexset empty during race reporting?

Yes, always.

> If it's not empty why we don't them to report? etc.

I've been using this example code:

```
#include <pthread.h>

#include <sanitizer/tsan_interface.h>

typedef unsigned long long invisible_barrier_t;

// Default instance of the barrier, but a test can declare more manually.
invisible_barrier_t barrier;

#ifdef __cplusplus
extern "C" {
#endif
void __tsan_testonly_barrier_init(invisible_barrier_t *barrier,
    unsigned count);
void __tsan_testonly_barrier_wait(invisible_barrier_t *barrier);
unsigned long __tsan_testonly_shadow_stack_current_size();
#ifdef __cplusplus
}
#endif

static inline void barrier_init(invisible_barrier_t *barrier, unsigned count) {
  __tsan_testonly_barrier_init(barrier, count);
}

static inline void barrier_wait(invisible_barrier_t *barrier) {
  __tsan_testonly_barrier_wait(barrier);
}

int Global;
pthread_mutex_t mtx;

void *Thread1(void *x) {
  barrier_wait(&barrier);
  pthread_mutex_lock(&mtx);
  Global++;
  pthread_mutex_unlock(&mtx);
  return NULL;
}

void *Thread2(void *x) {
  Global--;
  barrier_wait(&barrier);
  return NULL;
}

int main() {
  barrier_init(&barrier, 2);
  // CHECK: WARNING: ThreadSanitizer: data race
  // CHECK:   Write of size 4 at {{.*}} by thread T1
  // CHECK:                         (mutexes: write [[M1:M[0-9]+]]):
  // CHECK:   Previous write of size 4 at {{.*}} by thread T2:
  // CHECK:   Mutex [[M1]] (0x{{.*}}) created at:
  // CHECK:     #0 pthread_mutex_init
  // CHECK:     #1 main {{.*}}mutexset1.cc:[[@LINE+1]]
  pthread_mutex_init(&mtx, 0);
  pthread_t t[2];
  pthread_create(&t[0], NULL, Thread1, NULL);
  pthread_create(&t[1], NULL, Thread2, NULL);
  pthread_join(t[0], NULL);                                                                                                           
  pthread_join(t[1], NULL);
  pthread_mutex_destroy(&mtx);
}
```

The correct sequence of changes of the mutexset size before detecting the bug could help me to find the root cause... comparing with my local copy.

I will be back to it once I will finish the exit/join race.

**Comment by dvyukov** [comment]
> Yes. We call Add(), Remove(), Add(), Remove(). And when we report any issue, mutexset is always 0.

Then the question is: when/how it becomes empty between mutex lock and race report?..

**Comment by krytarowski** [comment]
OK, I think I know what's wrong. We call pthread_mutex_lock and pthread_mutex_unlock with a hard linking inside libpthread. Add() Del() sequence is from the internals of this librery during the bootstrap.

Code out of libpthread(3) uses the libc stubs `__libc_mutex_lock` and `__libc_mutex_unlock`.

https://nxr.netbsd.org/xref/src/lib/libpthread/pthread.h#343

```
    342 #define	pthread_mutex_init		__libc_mutex_init
    343 #define	pthread_mutex_lock		__libc_mutex_lock
    344 #define	pthread_mutex_trylock		__libc_mutex_trylock
    345 #define	pthread_mutex_unlock		__libc_mutex_unlock
    346 #define	pthread_mutex_destroy		__libc_mutex_destroy
    347 
    348 #define	pthread_mutexattr_init		__libc_mutexattr_init
    349 #define	pthread_mutexattr_settype	__libc_mutexattr_settype
    350 #define	pthread_mutexattr_destroy	__libc_mutexattr_destroy
```

There are strong aliases inside libpthread:

https://nxr.netbsd.org/xref/src/lib/libpthread/pthread_mutex.c#111

```
    110 __strong_alias(__libc_mutex_init,pthread_mutex_init)
    111 __strong_alias(__libc_mutex_lock,pthread_mutex_lock)
    112 __strong_alias(__libc_mutex_trylock,pthread_mutex_trylock)
    113 __strong_alias(__libc_mutex_unlock,pthread_mutex_unlock)
    114 __strong_alias(__libc_mutex_destroy,pthread_mutex_destroy)
    115 
    116 __strong_alias(__libc_mutexattr_init,pthread_mutexattr_init)
    117 __strong_alias(__libc_mutexattr_destroy,pthread_mutexattr_destroy)
    118 __strong_alias(__libc_mutexattr_settype,pthread_mutexattr_settype)
```

I've confirmed this in the assembly code:

```
000000000046e130 <_Z7Thread1Pv>:
  46e130:       55                      push   %rbp
  46e131:       48 89 e5                mov    %rsp,%rbp
  46e134:       48 83 ec 30             sub    $0x30,%rsp
  46e138:       48 8b 45 08             mov    0x8(%rbp),%rax
  46e13c:       48 89 7d f0             mov    %rdi,-0x10(%rbp)
  46e140:       48 89 c7                mov    %rax,%rdi
  46e143:       e8 e8 dc fd ff          callq  44be30 <__tsan_func_entry>
  46e148:       48 8b 45 f0             mov    -0x10(%rbp),%rax
  46e14c:       48 89 45 f8             mov    %rax,-0x8(%rbp)
  46e150:       b9 78 9a 4f 01          mov    $0x14f9a78,%ecx
  46e155:       89 cf                   mov    %ecx,%edi
  46e157:       e8 74 00 00 00          callq  46e1d0 <_ZL12barrier_waitPy>
  46e15c:       e9 00 00 00 00          jmpq   46e161 <_Z7Thread1Pv+0x31>
  46e161:       b8 88 9a 4f 01          mov    $0x14f9a88,%eax
  46e166:       89 c7                   mov    %eax,%edi
  46e168:       e8 f3 d5 f9 ff          callq  40b760 <__libc_mutex_lock@plt>
  46e16d:       89 45 ec                mov    %eax,-0x14(%rbp)
  46e170:       e9 00 00 00 00          jmpq   46e175 <_Z7Thread1Pv+0x45>
  46e175:       b8 80 9a 4f 01          mov    $0x14f9a80,%eax
  46e17a:       89 c7                   mov    %eax,%edi
  46e17c:       8b 05 fe b8 08 01       mov    0x108b8fe(%rip),%eax        # 14f9a80 <Global>
  46e182:       83 c0 01                add    $0x1,%eax
  46e185:       89 45 e8                mov    %eax,-0x18(%rbp)
  46e188:       e8 43 9b fd ff          callq  447cd0 <__tsan_write4>
  46e18d:       8b 45 e8                mov    -0x18(%rbp),%eax
  46e190:       89 05 ea b8 08 01       mov    %eax,0x108b8ea(%rip)        # 14f9a80 <Global>
  46e196:       b8 88 9a 4f 01          mov    $0x14f9a88,%eax
  46e19b:       89 c7                   mov    %eax,%edi
  46e19d:       e8 9e d4 f9 ff          callq  40b640 <__libc_mutex_unlock@plt>
  46e1a2:       89 45 e4                mov    %eax,-0x1c(%rbp)
  46e1a5:       e9 00 00 00 00          jmpq   46e1aa <_Z7Thread1Pv+0x7a>
  46e1aa:       e8 31 dd fd ff          callq  44bee0 <__tsan_func_exit>
  46e1af:       31 c0                   xor    %eax,%eax
  46e1b1:       48 83 c4 30             add    $0x30,%rsp
  46e1b5:       5d                      pop    %rbp
  46e1b6:       c3                      retq   
  46e1b7:       89 d1                   mov    %edx,%ecx
  46e1b9:       89 4d e0                mov    %ecx,-0x20(%rbp)
  46e1bc:       48 89 45 d8             mov    %rax,-0x28(%rbp)
  46e1c0:       e8 1b dd fd ff          callq  44bee0 <__tsan_func_exit>
  46e1c5:       48 8b 7d d8             mov    -0x28(%rbp),%rdi
  46e1c9:       e8 f0 25 00 00          callq  4707be <_Unwind_Resume>
  46e1ce:       66 90                   xchg   %ax,%ax
```

This means that I need to add alias name for the NetBSD case in TSan for every mutex-related function.

As far as I can tell, this name mangling was developed in multiple possible implementations in sight (and libpthread(3) that is dlopen(3)-able as required by some standards), but recently there is a attitude towards streamlining the names and remove this indirection.

**Comment by krytarowski** [comment]
It works!

```
$ ./mutex                                                                                                                                                                            
==================
WARNING: ThreadSanitizer: data race (pid=8877)
  Write of size 4 at 0x0000014f9a80 by thread T1 (mutexes: write M77):
    #0 Thread1(void*) /public/llvm-build/mutex.cc:35:9 (mutex+0x46e23c)

  Previous write of size 4 at 0x0000014f9a80 by thread T2:
    #0 Thread2(void*) /public/llvm-build/mutex.cc:41:9 (mutex+0x46e307)

  Location is global 'Global' of size 4 at 0x0000014f9a80 (mutex+0x0000014f9a80)

  Mutex M77 (0x0000014f9a88) created at:
    #0 pthread_mutex_lock /public/llvm/projects/compiler-rt/lib/tsan/../sanitizer_common/sanitizer_common_interceptors.inc:3801:3 (mutex+0x4278ae)
    #1 Thread1(void*) /public/llvm-build/mutex.cc:34:3 (mutex+0x46e21c)

  Thread T1 (tid=2, running) created by main thread at:
    #0 pthread_create /public/llvm/projects/compiler-rt/lib/tsan/rtl/tsan_interceptors.cc:921:3 (mutex+0x412840)
    #1 main /public/llvm-build/mutex.cc:57:3 (mutex+0x46e3a5)

  Thread T2 (tid=3, finished) created by main thread at:
    #0 pthread_create /public/llvm/projects/compiler-rt/lib/tsan/rtl/tsan_interceptors.cc:921:3 (mutex+0x412840)
    #1 main /public/llvm-build/mutex.cc:58:3 (mutex+0x46e3c4)

SUMMARY: ThreadSanitizer: data race /public/llvm-build/mutex.cc:35:9 in Thread1(void*)
==================
ThreadSanitizer: reported 1 warnings
```

**Comment by dvyukov** [comment]
Great!
Does it mean that we basically missed part of pthread_mutex_lock/unlock's? (which would be pretty bad to tsan)

**Comment by krytarowski** [comment]
I'm scanning the code to find more references of similar invisible functions. Once done I will submit a patch for review.

**Comment by krytarowski** [comment]
With the pending on review patches (name demangling and intercepting netbsd specific libpthread indirect symbol names in tsan) handling mutexes seems to work nicely.

http://www.netbsd.org/~kamil/llvm/tsan.2017-11-20.txt

On the other hand there is a fallout of breakage in certain functions like pthread_once(3) that deadlock now (we might need a fix similar to Apple). Once I will fix properly the thread-exit/join race I will squash the remaining bugs.

**Comment by krytarowski** [comment]
After finishing the pending patch for Thread Joined/Exited race.. time for setjmp(3)/longjmp(3).

https://github.com/freebsd/freebsd/blob/master/lib/libc/amd64/gen/setjmp.S
https://github.com/llvm-mirror/compiler-rt/blob/master/lib/tsan/rtl/tsan_rtl_amd64.S

Does TSan use a custom function for these libc calls?

I don't understand why FreeBSD has in libc e.g.:

```
movq	%rsp,16(%rcx)		/* 2; rsp */
```

and in TSan:

```
#if defined(__FreeBSD__)
  lea 8(%rsp), %rdi
  mov %rdi, %rsi
```

16 vs 8?

I need to port NetBSD specific values and I don't understand how to alter indices.

**Comment by dvyukov** [comment]
These are unrelated things.

`movq	%rsp,16(%rcx)` saves current rsp in jmp_buf at offset 16 (presumably that's where it is on freebsd).

`lea 8(%rsp), %rdi` obtains current rsp value to pass as argument to __tsan_setjmp.

It's still 8 vs 0. But I think in the end they end up being equal. Note tsan obtains rsp after `push %rdi`, but freebsd code does not (it has a push, but also a pop before saving rsp). I think that's the reason for 8 vs 0 difference.

There is relatively easy way to test/debug this. Add debug printfs to SetJmp/LongJmp in tsan_interceptors.cc and print sp in both cases. This statement must be true `if (buf->mangled_sp == mangled_sp)`. If it is not adjust offset in setjmp interceptor as necessary. Once they are equal, everything works.

**Comment by krytarowski** [comment]
I got the setjmp(3) family of functions to work.. however the patch is ugly. We need to deal with symbol mangling (for historical / compat reasons).

I will post it as is to review and discuss it there.

```
Failing Tests (9):
    ThreadSanitizer-x86_64 :: dtls.c
    ThreadSanitizer-x86_64 :: ignore_lib5.cc
    ThreadSanitizer-x86_64 :: ignored-interceptors-mmap.cc
    ThreadSanitizer-x86_64 :: mutex_lock_destroyed.cc
    ThreadSanitizer-x86_64 :: signal_errno.cc
    ThreadSanitizer-x86_64 :: signal_malloc.cc
    ThreadSanitizer-x86_64 :: signal_sync2.cc
    ThreadSanitizer-x86_64 :: signal_thread.cc
    ThreadSanitizer-x86_64 :: vfork.cc

  Expected Passes    : 286
  Expected Failures  : 1
  Unsupported Tests  : 83
  Unexpected Failures: 9
```

**Comment by krytarowski** [comment]
```
$ cat tt2.c
#include <pthread.h>
#include <stdio.h>
#include <stdlib.h>
#include <errno.h>

int main(int argc, char *argv[]) { 
  int ret;

  pthread_mutex_t m;
  pthread_mutex_init(&m, 0);
  pthread_mutex_lock(&m);
  pthread_mutex_unlock(&m);
  pthread_mutex_destroy(&m);

  ret = pthread_mutex_lock(&m);

  printf("ret=%d errno=%d\n", ret, errno);

  return 0;
}
```

This program tests UB. It's used in `mutex_lock_destroyed.cc`.

Linux, FreeBSD return EINVAL and errno 0.
NetBSD, OpenBSD return 0 and errno 0.

This shouldn't base on implementation-defined behavior in TSan.

I will try to implement proper check inside sanitizers.

**Comment by dvyukov** [comment]
Indeed UB. Will it help to create the mutex with PTHREAD_MUTEX_ERRORCHECK type?

**Comment by krytarowski** [comment]
PTHREAD_MUTEX_ERRORCHECK is a GNU extension.

On NetBSD there is PTHREAD_DIAGASSERT, but it does not catch this use-case, at least as of today.

**Comment by krytarowski** [comment]
There are still few unexpected failures in tests.

```
********************
Testing: 0 .. 10.. 20.. 30.. 40.. 50.. 60.. 70.. 80.. 90.. 
Testing Time: 50.74s
********************
Failing Tests (7):
    ThreadSanitizer-x86_64 :: deadlock_detector_stress_test.cc
    ThreadSanitizer-x86_64 :: dtls.c
    ThreadSanitizer-x86_64 :: ignore_lib5.cc
    ThreadSanitizer-x86_64 :: ignored-interceptors-mmap.cc
    ThreadSanitizer-x86_64 :: mutex_lock_destroyed.cc
    ThreadSanitizer-x86_64 :: signal_block.cc
    ThreadSanitizer-x86_64 :: vfork.cc

  Expected Passes    : 288
  Expected Failures  : 1
  Unsupported Tests  : 83
  Unexpected Failures: 7
```

signal_block.cc - there is some race, it sometimes works sometimes it does not work.
dtls.c - it looks like dynamically allocated TLS vectors missing on the NetBSD side
vfork.cc - testing UB, it looks like NetBSD behaves the same way like Linux, however the test is failing
mutex_lock_destroyed.cc - it bases on UB implemented in the Linux way

The other tests fail for similar narrow-case scenarios like massive mmap(2) calls that seem to overflow the shadow.

I've decided to stop now, focus on upstreaming local patches and move on to MSan (who is responsible for MSan?).

There are 4 pending patches in review and 2 still kept locally:
 - defer the start of a background thread,
 - enable tsan in cmake.

I plan to return to failures in sanitizers once I will finish LLDB/NetBSD porting.

Thank you for great help so far, you have made TSan/NetBSD possible!

Closing this thread as it is and back to LLVM Phabricator for the remaining patches.
